pkg/providers/vsphere/reconciler/reconciler.go

```diff
@@ -93,201 +71,74 @@ func SetupEnvVars(ctx context.Context, vsphereDatacenter *anywherev1.VSphereData
-		values["controlPlaneTemplateName"] = common.CPMachineTemplateName(clusterName, time.Now)
-		controlPlaneUser := machineConfigMap[cluster.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users[0]
-		values["vsphereControlPlaneSshAuthorizedKey"] = controlPlaneUser.SshAuthorizedKeys[0]
-
-		if cluster.Spec.ExternalEtcdConfiguration != nil {
-			etcdUser := machineConfigMap[cluster.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users[0]
-			values["vsphereEtcdSshAuthorizedKey"] = etcdUser.SshAuthorizedKeys[0]
-		}
-
-		values["etcdTemplateName"] = common.EtcdMachineTemplateName(clusterName, time.Now)
-	}
-	log.Info("cluster", "name", cluster.Name)
-
-	if result, err := r.reconcileControlPlaneSpec(ctx, cluster, templateBuilder, specWithBundles, cpOpt, log); err != nil {
-		return result, err
-	}
+	return controller.NewPhaseRunner().Register(
+		r.ValidateDatacenterConfig,
+		r.ValidateMachineConfigs,
+		r.ReconcileControlPlane,
+		r.ReconcileCNI,
+		r.ReconcileWorkers,
+	).Run(ctx, log, clusterSpec)
+}
 
-	if result, err := r.reconcileWorkerNodeSpec(ctx, cluster, templateBuilder, specWithBundles, workloadTemplateNames, kubeadmconfigTemplateNames); err != nil {
-		return result, err
-	}
+func (r *Reconciler) ValidateDatacenterConfig(ctx context.Context, log logr.Logger, clusterSpec *c.Spec) (controller.Result, error) {
+	log = log.WithValues("phase", "validateDatacenterConfig")
+	dataCenterConfig := clusterSpec.VSphereDatacenter
 
-	capiCluster, result, errCAPICLuster := r.getCAPICluster(ctx, cluster, log)
-	if errCAPICLuster != nil {
-		return result, errCAPICLuster
+	// Set up envs for executing Govc cmd and default values for datacenter config
+	if err := SetupEnvVars(ctx, dataCenterConfig, r.client); err != nil {
+		log.Error(err, "Failed to set up env vars and default values for VSphereDatacenterConfig")
+		return controller.Result{}, err
 	}
-
-	// wait for etcd if necessary
-	if cluster.Spec.ExternalEtcdConfiguration != nil {
-		if !conditions.Has(capiCluster, managedEtcdReadyCondition) || conditions.IsFalse(capiCluster, managedEtcdReadyCondition) {
-			log.Info("Waiting for etcd to be ready", "cluster", cluster.Name)
-			return controller.Result{Result: &ctrl.Result{
-				RequeueAfter: defaultRequeueTime,
-			}}, nil
+	if !dataCenterConfig.Status.SpecValid {
+		failureMessage := fmt.Sprintf("VSphereDatacenterConfig %s is invalid", dataCenterConfig.Name)
+		if dataCenterConfig.Status.FailureMessage != nil {
+			failureMessage += ": " + *dataCenterConfig.Status.FailureMessage
 		}
-	}
-
-	if !conditions.IsTrue(capiCluster, controlPlaneReadyCondition) {
-		log.Info("waiting for control plane to be ready", "cluster", capiCluster.Name, "kind", capiCluster.Kind)
-		return controller.Result{Result: &ctrl.Result{
-			RequeueAfter: defaultRequeueTime,
-		}}, err
-	}
 
-	if result, err := r.reconcileCNI(ctx, cluster, capiCluster, specWithBundles, log); err != nil {
-		return result, err
+		log.Error(nil, failureMessage)
```

Comment:
maybe include the VSphereDatacenterConfig as a value in the log

Reply:
I would prefer to keep the messaging for this as it currently is

Reply:
Ok. Why?
If we don't add the datacenter, the message doesn't tell us which datacenter is invalid

```go
log.Error(errors.New(failureMessage), "Invalid VSphereDatacenterConfig, "datacenterConfig", klog.KObj(dataCenterConfig))
```

Reply:
Oh we're adding the datacenterConfig name to the failure message a few lines above: https://github.com/aws/eks-anywhere/blob/72b11b3e59aa801154dbc554e063b8fdd7de761c/pkg/providers/vsphere/reconciler/reconciler.go#L94

But now i see your idea of just adding it as a log value. That works for me too!